src/pip/_internal/resolution/resolvelib/candidates.py

```diff
@@ -120,7 +127,46 @@ def dist(self):
 
     def get_dependencies(self):
         # type: () -> Sequence[InstallRequirement]
-        return [
-            self._make_install_req(str(r), self._ireq)
-            for r in self.dist.requires()
+        return [self._make_install_req(str(r)) for r in self.dist.requires()]
+
+    def get_install_requirement(self):
+        # type: () -> InstallRequirement
+        return self._ireq
+
+
+class ExtrasCandidate(LinkCandidate):
+    def __init__(
+        self,
+        base,      # type: LinkCandidate
+        extras,    # type: Set[str]
+    ):
+        # type: (...) -> None
+        self.base = base
+        self.extras = extras
+        self.link = base.link
+
+    @property
+    def name(self):
+        # type: () -> str
+        """The normalised name of the project the candidate refers to"""
+        return format_name(self.base.name, self.extras)
+
+    @property
+    def version(self):
+        # type: () -> _BaseVersion
+        return self.base.version
+
+    def get_dependencies(self):
+        # type: () -> Sequence[InstallRequirement]
+        deps = [
+            self.base._make_install_req(str(r))
+            for r in self.base.dist.requires(self.extras)
         ]
+        # Add a dependency on the exact base
+        spec = "{}=={}".format(self.base.name, self.base.version)
+        deps.append(self.base._make_install_req(spec))
+        return deps
```

Comment:
Probably needs a comment or docstring explianing the implementation here for future maintainers 😛

Reply:
So you're not a fan of the "if it was hard to write, it should be hard to understand" philosophy then? 😉 

I'll add a comment here.

Reply:
Added a docstring (more like an essay really 😉)